gh-102468: Document `PyCFunction_New*` and `PyCMethod_New` (#112557)
python/cpython: Doc/c-api/structures.rst

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:function:: PyObject * PyCFunction_New(PyMethodDef *ml, PyObject *self)

>    Turn ``ml`` into a Python callable object.  The ``self`` parameter will be

Reviewer: `ml` and `self` should be marked up in italic (see the devguide's style guide section for details). Please apply to the whole PR.

```suggestion
   Turn *ml* into a Python callable object.  The *self* parameter will be
```

   passed as ``self`` parameter to the C function in ``ml->ml_meth`` when
   invoked, can be ``NULL``.

.. c:function:: PyObject * PyCFunction_NewEx(PyMethodDef *ml, PyObject *self, PyObject *module)

   Same as :c:func:`PyCFunction_New`, but also allows setting the function

>    object's ``__module__`` attribute.  ``module`` can be anything.

Reviewer: For `__module__`, I think you can link to the reference docs. Check the reference to find the correct Sphinx anchor.

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:function:: PyObject * PyCFunction_New(PyMethodDef *ml, PyObject *self)

>    Turn *ml* into a Python :term:`callable`.  The *self* parameter will be

Reviewer: I think we should throw "object" back into this sentence; it reads better.

```suggestion
   Turn *ml* into a Python :term:`callable` object.  The *self* parameter will be
```

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:function:: PyObject * PyCMethod_New(PyMethodDef *ml, PyObject *self, PyObject *module, PyTypeObject *cls)

>    Turn *ml* into a Python :term:`callable` object.  The caller must ensure that
>    *ml* outlives the :term:`callable`.  Typically, *ml* is defined as a static
>    variable.

Reviewer: SemBr:

```suggestion
   Turn *ml* into a Python :term:`callable` object.
   The caller must ensure that *ml* outlives the :term:`callable`.
   Typically, *ml* is defined as a static variable.
```

>    The *self* parameter will be passed as the *self* argument to the C function
>    in ``ml->ml_meth`` when invoked.  *self* can be ``NULL``.
> 
>    The :term:`callable` object's ``__module__`` attribute can be set from the
>    given *module* argument.  *module* should be a Python string, which will be
>    used as name of the module the function is defined in.  If unavailable, it
>    can be set to :const:`None` or ``NULL``.

Reviewer: SemBr:

```suggestion
   The *self* parameter will be passed as the *self* argument
   to the C function in ``ml->ml_meth`` when invoked.
   *self* can be ``NULL``.

   The :term:`callable` object's ``__module__`` attribute
   can be set from the given *module* argument.
   *module* should be a Python string,
   which will be used as name of the module the function is defined in.
   If unavailable, it can be set to :const:`None` or ``NULL``.
```

>    The *cls* parameter will be passed as the *defining_class* argument to the C
>    function.  Must be set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.

Reviewer: SemBr:

```suggestion
   The *cls* parameter will be passed as the *defining_class*
   argument to the C function.
   Must be set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.
```

>    .. seealso: :attr:`function.__module__`

Reviewer 2: Please move this just below the *module* argument doc.

Author: The seealso not rendered at https://cpython-previews--112557.org.readthedocs.build/en/112557/c-api/structures.html#c.PyCMethod_New , maybe it must at the end of a section?

Author: Ah, I got it, there is a missing colon.

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:function:: PyObject * PyCFunction_New(PyMethodDef *ml, PyObject *self)

   Turn *ml* into a Python callable object.  The *self* parameter will be

>    passed as *self* parameter to the C function in ``ml->ml_meth`` when
>    invoked, can be ``NULL``.

Reviewer: ```suggestion
   passed as the *self* argument to the C function in ``ml->ml_meth`` when
   invoked. *self* can be ``NULL``.
```

   .. note::

      This function "steals" a reference to *ml*.

.. c:function:: PyObject * PyCFunction_NewEx(PyMethodDef *ml, PyObject *self, PyObject *module)

>    Same as :c:func:`PyCFunction_New`, but also allows setting the function
>    object's ``__module__`` attribute.  :attr:`!__module__` should be the name
>    of the module the function is defined in or ``None`` if unavailable.
>    See :attr:`function.__module__`.

Reviewer: ```suggestion
   Same as :c:func:`PyCFunction_New`, but also allows setting the function
   object's ``__module__`` attribute from the given *module* argument.
   *module* should be the name of the module the function is defined in,
   or ``None`` if unavailable.

   .. seealso: :attr:`function.__module__`
```

   .. note::

      This function "steals" a reference to *ml*.

.. c:function:: PyObject * PyCMethod_New(PyMethodDef *ml, PyObject *self, PyObject *module, PyTypeObject *cls)

>    Same as :c:func:`PyCFunction_NewEx`, but accept a **cls** parameter, which
>    will be passed as ``defining_class`` parameter to the C function.  Must be
>    set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.

Reviewer: ```suggestion
   Same as :c:func:`PyCFunction_NewEx`, but accept a *cls* parameter, which
   will be passed as the *defining_class* argument to the C function.
   Must be set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.
```

I think the last sentence reads a little bit weird; what must be set? The *cls* parameter?

   slot.  This is helpful because calls to PyCFunctions are optimized more
   than wrapper object calls.

.. c:function:: PyObject * PyCFunction_New(PyMethodDef *ml, PyObject *self)

>    Turn *ml* into a Python callable object.  The *self* parameter will be

Reviewer: ```suggestion
   Turn *ml* into a Python :term:`callable`.  The *self* parameter will be
```

   passed as the *self* argument to the C function in ``ml->ml_meth`` when
   invoked. *self* can be ``NULL``.

> 
>    .. note::
> 
>       This function "steals" a reference to *ml*.

Author: ```suggestion
```

.. c:function:: PyObject * PyCFunction_NewEx(PyMethodDef *ml, PyObject *self, PyObject *module)

   Same as :c:func:`PyCFunction_New`, but also allows setting the function
   object's ``__module__`` attribute from the given *module* argument.
   *module* should be the name of the module the function is defined in,
   or ``None`` if unavailable.

   .. seealso: :attr:`function.__module__`

> 
>    .. note::
> 
>       This function "steals" a reference to *ml*.

Author: ```suggestion
```

.. c:function:: PyObject * PyCMethod_New(PyMethodDef *ml, PyObject *self, PyObject *module, PyTypeObject *cls)

   Same as :c:func:`PyCFunction_NewEx`, but accept a *cls* parameter, which
   will be passed as the *defining_class* argument to the C function.
   Must be set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.

> 
>    .. note::
> 
>       This function "steals" a reference to *ml*.

Author: ```suggestion
```

   Turn *ml* into a Python :term:`callable` object.
   The caller must ensure that *ml* outlives the :term:`callable`.
   Typically, *ml* is defined as a static variable.

   The *self* parameter will be passed as the *self* argument
   to the C function in ``ml->ml_meth`` when invoked.
   *self* can be ``NULL``.

   The :term:`callable` object's ``__module__`` attribute
   can be set from the given *module* argument.
   *module* should be a Python string,
   which will be used as name of the module the function is defined in.
   If unavailable, it can be set to :const:`None` or ``NULL``.

   .. seealso: :attr:`function.__module__`

   The *cls* parameter will be passed as the *defining_class*
   argument to the C function.
   Must be set if :c:macro:`METH_METHOD` is set on ``ml->ml_flags``.

.. c:function:: PyObject * PyCFunction_NewEx(PyMethodDef *ml, PyObject *self, PyObject *module)

   Equivalent to ``PyCMethod_New(ml, self, module, NULL)``.

.. c:function:: PyObject * PyCFunction_New(PyMethodDef *ml, PyObject *self)

   Equivalent to ``PyCMethod_New(ml, self, NULL, NULL)``.

>    .. versionadded:: 3.9

Reviewer 2: `PyCMethod_New` was added in 3.9, not `PyCFunction_New`

Author: Sorry for the mistake, updated!
